consensus/core/src/synchronizer.rs

```diff
@@ -649,6 +692,111 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                                    for serialized_block in result {
+                                        let Ok(signed_block) = bcs::from_bytes(&serialized_block) else {
+                                            warn!("Malformed block received from peer {authority_index} while fetching own last block, skipping.");
+                                            continue 'main_loop;
+                                        };
+
+                                        if let Err(e) = block_verifier.verify(&signed_block) {
+                                            context
+                                                .metrics
+                                                .node_metrics
+                                                .invalid_blocks
+                                                .with_label_values(&[&signed_block.author().to_string(), "synchronizer_own_block"])
+                                                .inc();
+                                            warn!("Invalid block received from {}: {}", authority_index, e);
+                                            continue 'main_loop;
+                                        }
+
+                                        let verified_block = VerifiedBlock::new_verified(signed_block, serialized_block);
+                                        if verified_block.author() != context.own_index {
+                                            continue 'main_loop;
+                                        }
+                                        blocks.push(verified_block);
+                                    }
+
+                                    let max_round = blocks.into_iter().map(|b|b.round()).max().unwrap_or(0);
+                                    highest_round = highest_round.max(max_round);
+
+                                    total_stake += context.committee.stake(authority_index);
+
+                                    if results.is_empty() {
+                                        break;
+                                    }
+                                },
+                                Err(err) => {
+                                    warn!("Error {err} while fetching our own block from peer {authority_index}. Will retry.");
+                                    results.push(fetch_own_block(authority_index, FETCH_OWN_BLOCK_RETRY_DELAY));
+                                }
+                            }
+                        },
+                        () = &mut timer => {
+                            info!("Timeout while trying to sync our own last block from peers");
+                            break;
+                        }
+                    }
+                }
+
+                // Update the Core with the highest detected round
+                if total_stake == 0 {
+                    panic!("No peer has returned any acceptable result, can not safely update min round");
```

Comment:
Panic is fine. But I wonder if it is better to always retry until receiving at least f+1 blocks.

Reply:
Yeah I believe the switch to the retry will be kind of inevitable if we refactor the amnesia recovery to run automatically. We can always run to the (well really) edge case where the network attempts to recover after a switch to new epoch (where all nodes crashed for whatever reason) and at this point we can't afford to have nodes panicing after a few seconds when they don't successfully get back results due to staggered boot. I guess that's a case might come up mostly in simtests etc